### PR TITLE
Fix/enable mode module types (m06)

### PR DESCRIPTION
**Pull request by filmakarov**
Fix Issue #[30](https://cantina.xyz/code/d1d4b139-9705-4367-9468-297b7078674e/findings/30)) : Enable mode would not install any module types but validators or multitypes with validator included.

Also includes fixes to issues 37,38,42 from [PR#111](https://github.com/bcnmy/nexus/pull/111)

### CODE REVIEW

File: contracts/Nexus.sol [hunk 1]
```diff
@@ -100,14 +100,15 @@ contract Nexus is INexus, BaseAccount, ExecutionHelper, ModuleManager, UUPSUpgra
         uint256 missingAccountFunds
     ) external virtual payPrefund(missingAccountFunds) onlyEntryPoint returns (uint256 validationData) {
         address validator = op.nonce.getValidator();
-        if (!op.nonce.isModuleEnableMode()) {
+        if (op.nonce.isModuleEnableMode()) {
+            PackedUserOperation memory userOp = op;
+            userOp.signature = _enableMode(userOpHash, op.signature);
+            if (!_isValidatorInstalled(validator)) return VALIDATION_FAILED;
+            validationData = IValidator(validator).validateUserOp(userOp, userOpHash);
+        } else {
```
Comment (Aboudjem) [inline contracts/Nexus.sol:108]: Is there any reason the order changed? I think it's less likely to happen when calling with EnableMode enabled compared to when calling without EnableMode

File: contracts/types/Constants.sol [hunk 2]
```diff
@@ -38,7 +38,7 @@ uint256 constant MODULE_TYPE_FALLBACK = 3;
 
 // Module type identifier for hooks
 uint256 constant MODULE_TYPE_HOOK = 4;
-bytes32 constant MODULE_ENABLE_MODE_TYPE_HASH = keccak256("ModuleEnableMode(address module, bytes32 initDataHash)");
+bytes32 constant MODULE_ENABLE_MODE_TYPE_HASH = keccak256("ModuleEnableMode(address module, uint256 moduleType, bytes32 userOpHash, bytes32 initDataHash)");
```
Comment (Aboudjem) [inline contracts/types/Constants.sol:41]: According to the [[EIP-712 standard](https://eips.ethereum.org/EIPS/eip-712)](https://eips.ethereum.org/EIPS/eip-712), members are joined by a "," without any blank space.

`The type of a struct is encoded as name ‖ "(" ‖ member₁ ‖ "," ‖ member₂ ‖ "," ‖ … ‖ memberₙ ")" where each member is written as type ‖ " " ‖ name. For example, the above Mail struct is encoded as Mail(address from,address to,string contents).`

Check M-06 for more info.

Comment (livingrockrises) [reply contracts/types/Constants.sol:41]: @filmakarov

File: contracts/Nexus.sol (hunk 1 above)
Comment (livingrockrises) [reply contracts/Nexus.sol:108]: @filmakarov

Comment (filmakarov) [reply contracts/Nexus.sol:108]: https://cantina.xyz/code/d1d4b139-9705-4367-9468-297b7078674e/findings/42

File: contracts/types/Constants.sol (hunk 2 above)
Comment (filmakarov) [reply contracts/types/Constants.sol]: fixed!

**Review by livingrockrises — APPROVED** [review]

**Pull request merged**